### ISSUE TITLE
Bad memory corruption in shadow code, breaks everything

### ISSUE THREAD

**Issue by Paril** [jdolan/quetoo#270]
Related to @jdolan commits 217aeedf13982c20fb49245d01ce7902e6901221 and bf8d5f86fc7f4a5aca91283d237fd12a58b92ed7, one of them introduces the crash but not sure which or where.

Basically, the issue happens in R_UpdateShadows. l->shadows[i] 0 and 1 (sun and ambient?) are always corrupt, including a corrupt cm_bsp_plane_t which references huge plane numbers (like USHRT_MAX). This, in turn, corrupts the memory header/footer, and causes explosions in random systems that are alongside this block of memory. For me, it completely wipes surface memory (it randomly clears r_material_t pointers/instances sometimes, sets first_index to weird values, all sorts of crazy shit), and causes crashes on quit/restart because the linked list in memory blocks are destroyed.

Cause unknown as of yet. Happens with threads -1 as well. r_shadows 0 fixes the issue, as does undoing the two commits listed.

**Comment by jdolan** [comment]
Wow! Interesting. So.. let me explain what `r_shadow_state.plane_shadow_counts` and `plane->num` are for.

Quake2 has "duplicates" of each plane -- front and back facing, essentially. This tripped me up -for days- when implementing the shadow code the first time, because when casting shadows on north and east facing walls, the collision code would return one "side" of the plane, and on south and west facing walls, it would return the other. The problem? BSP surfaces *almost never* reference the odd numbered ("back") planes. Instead, for some insane reason, BSP surfaces reference the even numbered ("front") plane, and use a separate bit-mask for `R_SURF_PLANE_BACK`. So I was using the back plane as my shadow plane, and writing its number to the stencil buffer, but shadows would never pass the stencil test for those planes. Basically, shadows were never cast on walls that faced certain directions.. I lost a week to this.

Here, this is worth a laugh:
https://github.com/jdolan/quetoo/commit/3bd5ffd12d8383953b67b6605587343d4fe666b0

Ultimately, when I realized what was happening, I came up with the "trick" for writing the correct value to the stencil buffer: `(plane->num >> 1) % 255 + 1`. Halving the value ensures both sides of each plane produce the same value. Modulo 255 ensures the value fits in an 8 bit stencil buffer. And +1 ensures we're writing values `1-255`, and never `0`.

***

As for `r_shadow_state->plane_surface_counts`, this is meant to hold a count of the number of shadows referencing a given plane. This was implemented as a fix for #237. At first, I had tried to simply add a shadow count on the plane struct, but modifying this value from the entity culling thread while the renderer was drawing the world was leading to crashes -- *I chalked this up to optimizations the compiler was making around `const`, but maybe that is not the case and something is in fact wrong with my writing to this field??* So I moved it out to its own array, external to the world model entirely. I'm surprised that there are still crashes here.. but looking into this now.
